### ISSUE TITLE
[Bug] cmder loads very slowly under Windows

### ISSUE THREAD

**Issue by minstyle** [cmderdev/cmder#2816]
### Version Information

```markdown
Cmder version: v1.3.21 (1.3.20.151)
Clink: v1.4.11
git: version 2.39.1.windows.1
ConEmu: 221218 [64] Stable

Operating system: Windows 10 Pro
```

### Cmder Edition

Cmder Full (with Git)

### Description of the issue

Since the latest Windows 10 update 22H2, it takes 30 seconds or more for cmder to load. It also takes this long when creating new tab windows in cmder. The problem exists with different Windows 10 computers at different work spaces and locations (private, work and Home Office). 

### Checklist

- [X] I have read the documentation.
- [X] I have searched for similar issues and found none that describe my issue.
- [X] I have reproduced the issue on the latest version of Cmder.
- [X] I am certain my issues are not related to ConEmu, Clink, or other third-party tools that Cmder uses.

**Comment by mpitz** [comment]
Same here!

```
Cmder version: v1.3.21 (1.3.20.151)
Clink: v1.4.10.45c041
git: version 2.39.0.windows.1
ConEmu: 221218 [64] Stable

Operating system: Windows 11 Pro 22H2 (Build 22621.1105)
```

**Comment by Lari0** [comment]
Hi, 
Same here , it takes 20 or more seconds with /f  and close to 60 without /f .
Also tried adding variable of git location , this improves startup for few seconds.

Elapsed Time: 0:0:22.87 (22.87s total)

Microsoft Windows [Version 10.0.19045.2251]  - Windows 10 Pro 22H2
cmder 1.3.20.151

**Comment by chrisant996** [comment]
I am using Cmder 1.3.21 Mini version, with Windows 10 Pro 22H2 19045.2486, and it takes about 2 seconds to start.

I have git for Windows installed, and Cmder finds it successfully.

**Comment by daxgames** [comment]
I can not duplicate this.

**Comment by WickedSilver** [comment]
> I can not duplicate this.

Is there any thing we can do to help you to reproduce this?

**Comment by mpitz** [comment]
Don't know if it helps but starting Cmder I can see rotating cmd.exe and findstr.exe in the lower tab:

![image](https://user-images.githubusercontent.com/6831223/213456067-a37d6a0c-f11b-4666-a1d2-b645e0ab1e95.png)

Happens on my Windows 10 Pro private PC an my Windows 11 Pro office PC. On the office PC the startup take significant more time than on the Win 10 machine.

- Win 10: 10 sec
- Win 11: 30 sec

**Comment by mpitz** [comment]
@daxgames maybe, but my work PCs has no extra firewall restrictions, I'am admin user. It's like a private PC.

**Comment by daxgames** [comment]
When I said I can not reproduce this I was in a hurry.  The only Windows 10 I have tested on is version 10.0.19044 21H2.2486 and according to Windows Ulpdate is up to date.

I am running Windows 10 on a 8GB  Virtualbox VM on an older I5 laptop with slow spindle HDD and even on this older system Cmder starts in 3.68 seconds.

I will try and find another system in my stack of old hardware and see what I have to try and reproduce but if I can't get the version you guys say the issue started at I'm not sure what to do.

**Comment by daxgames** [comment]
@mpitz I'm just telling you historical experience.  It comes up a lot.

**Comment by js-compilatrum** [comment]
Version mini and full at Microsoft Windows [Version 10.0.22621.1105] is very slow. It was checked on 128 GB RAM, 13700 CPU so hardware should will not be issue here. Older version on laptop is extremaly fast in comparision.

**Comment by chrisant996** [comment]
Here are some ideas to try, to learn more about where the delay is coming from:

- Try using Task Manager or Process Explorer to observe what processes are using CPU during the 10-30 second delay.  For example, it might show that some antivirus system is spending a lot of CPU.
- Try `set PATH` to check what the system path looks like.  For example, if it has UNC paths in the system PATH then that will make network calls when searching to find executables, and that could introduce delays.  Or there might be extra unexpected paths in the PATH that might slow things down.

**Comment by Dzoukr** [comment]
> Don't know if it helps but starting Cmder I can see rotating cmd.exe and findstr.exe in the lower tab:
> 
> ![image](https://user-images.githubusercontent.com/6831223/213456067-a37d6a0c-f11b-4666-a1d2-b645e0ab1e95.png)
> 
> Happens on my Windows 10 Pro private PC and my Windows 11 Pro office PC. On the office PC the startup takes significant more time than on the Win 10 machine.
> 
> * Win 10: 10 sec
> * Win 11: 30 sec

Absolutely the same here for me on Win 10 with cmder mini. I also tried to observe Task manager, but nothing significant there... Any idea what could cause it?

**Comment by chrisant996** [comment]
What does `set PATH` report?  For this, please copy/paste the actual text, not a screenshot (screenshots are images, so the text is not accessible, and trying to transcribe it into actual text is a time-consuming and error-prone process).

I ask because I multiple `findstr` calls in the `.bat` and `.cmd` scripts that are part of Cmder, and many of them involve trying to analyze the %PATH% variable.  It could potentially be possible for certain content in the %PATH% to lead to various problems, so examining the %PATH% can be a constructive diagnostic step.

**Comment by andrej-dyck** [comment]
Same. I extracted cmder.zip on a fresh install of Win10 22H2, and it takes up to 50sec to start.

**Comment by daxgames** [comment]
1. Found a system and got Windows 10 22H2 installed.
2. Downloaded [Cmder Full 1.3.21](https://github.com/cmderdev/cmder/releases/download/v1.3.21/cmder.zip) from the `cmder.app` web site.
    - The version file in the archive says it is `1.3.20.151` and not `1.3.21`.  See #2818.
4. I unziped and ran `cmder.exe`.
    - Launch speed seemed normal to me for a first launch.
    - Git for windows did its initial install thing.
5. I added `/t` to the `cmd::Cmder` task, saved and relaunched.  Had to do this twice for some reason???
6. Cmder started in less than 2 seconds.
    ![image](https://user-images.githubusercontent.com/7318053/213923962-cf4ab470-cb10-4f11-95c3-7b0f81062220.png)
7. Hardware is an older Intel i7 with 512Gb SSD and 16 GB RAM.

I cannot reproduce this issue.

I would love to see the output of path commands as requested above.

**Comment by andrej-dyck** [comment]
Here's some debug-output:
```
DEBUG(init.bat): Env Var - CMDER_ROOT=C:\<redacted-path>\Cmder

DEBUG(init.bat): Env Var - debug_output=1

DEBUG(init.bat): Looking for Git install root...

DEBUG(:read_version): Env Var - git_executable=C:\<redacted-path>\Cmder\vendor\git-for-windows\cmd\git.exe

DEBUG(:read_version): Env Var - GIT_VERSION_VENDORED=2.39.0.windows.1

DEBUG(:validate_version): ARGV[1]=VENDORED, ARGV[2]=2.39.0.windows.1

DEBUG(:parse_version): ARGV[1]=VENDORED, ARGV[2]=2.39.0.windows.1

DEBUG(:validate_version): Found Git Version for VENDORED: 2.39.0.windows.1

DEBUG(:read_version): Env Var - git_executable=C:\<redacted-path>\git-2.39.0\cmd\git.exe

DEBUG(:read_version): Env Var - GIT_VERSION_USER=2.39.0.windows.2

DEBUG(:get_user_git_version): get_user_git_version GIT_VERSION_USER: 2.39.0.windows.2

DEBUG(:validate_version): ARGV[1]=USER, ARGV[2]=2.39.0.windows.2

DEBUG(:parse_version): ARGV[1]=USER, ARGV[2]=2.39.0.windows.2

DEBUG(:validate_version): Found Git Version for USER: 2.39.0.windows.2

DEBUG(:compare_versions): Comparing:

DEBUG(:compare_versions): USER: 2.39.0.windows.2

DEBUG(:compare_versions): VENDORED: 2.39.0.windows.1

DEBUG(:compare_git_versions): campare versions_result: 1

DEBUG(init.bat): Using found Git '2.39.0.windows.2' from 'C:\<redacted-path>\git-2.39.0...

DEBUG(init.bat): Using Git from 'C:\<redacted-path>\git-2.39.0...

DEBUG(:enhance_path): Env Var INSIDE PATH C:\\<redacted-path>\\git-2.39.0\\cmd - found=1

DEBUG(:enhance_path): Env Var BEGIN PATH C:\\<redacted-path>\\git-2.39.0\\cmd - found=1

DEBUG(:enhance_path): Env Var INSIDE PATH C:\\<redacted-path>\\git-2.39.0\\mingw64\\bin - found=1

DEBUG(:enhance_path): Env Var END PATH C:\\<redacted-path>\\git-2.39.0\\mingw64\\bin - found=1

DEBUG(:enhance_path): Env Var INSIDE PATH C:\\<redacted-path>\\git-2.39.0\\usr\\bin - found=1

DEBUG(:enhance_path): Env Var END PATH C:\\<redacted-path>\\git-2.39.0\\usr\\bin - found=1

DEBUG(init.bat): Env Var - GIT_INSTALL_ROOT=C:\<redacted-path>\git-2.39.0

DEBUG(init.bat): Found Git in: 'C:\<redacted-path>\git-2.39.0'

DEBUG(:enhance_path): Env Var INSIDE PATH C:\\<redacted-path>\\Cmder\\vendor\\bin - found=0

DEBUG(:enhance_path): Env Var END PATH C:\\<redacted-path>\\Cmder\\vendor\\bin - found=0

DEBUG(:enhance_path): Appending 'C:\<redacted-path>\Cmder\vendor\bin'

DEBUG(:enhance_path): END Env Var - PATH=C:\<redacted-path>\Cmder\vendor\conemu-maximus5\ConEmu\Scripts;C:\<redacted-path>\Cmder\vendor\conemu-maximus5;C:\<redacted-path>\Cmder\vendor\conemu-maximus5\ConEmu;C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\wbin;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Windows\System32\OpenSSH\;C:\Program Files\Intel\WiFi\bin\;C:\Program Files\Common Files\Intel\WirelessCommon\;C:\Program Files\dotnet\;C:\Program Files\Docker\Docker\resources\bin;;C:\<redacted-path>\git-2.39.0\bin;C:\<redacted-path>\git-2.39.0\cmd;C:\<redacted-path>\git-2.39.0\mingw64\bin;C:\<redacted-path>\git-2.39.0\usr\bin;C:\<redacted-path>\Gradle\gradle-7.6\bin;C:\<redacted-path>\Haskell\ghc-8.10.4\bin;C:\<redacted-path>\Java\jdk-17.0.2\bin;C:\<redacted-path>\JetBrains\scripts;C:\<redacted-path>\Kubernetes\bin;C:\<redacted-path>\Make\bin;C:\<redacted-path>\Maven\apache-maven-3.6.3\bin;C:\<redacted-path>\Node\node-18.13.0;C:\<redacted-path>\PHP\php-8.1.3;C:\<redacted-path>\PHP\composer;C:\<redacted-path>\Python\python-3.8.1;C:\<redacted-path>\Terraform;;C:\Users\<redacted>\AppData\Local\Microsoft\WindowsApps;C:\Users\<redacted>\.dotnet\tools;C:\<redacted-path>\Cmder\vendor\bin

DEBUG(:enhance_path): Env Var C:\\<redacted-path>\\Cmder\\vendor\\bin - found=1

DEBUG(:enhance_path_recursive): Env Var - add_path=

DEBUG(:enhance_path_recursive): Env Var - position=append

DEBUG(:enhance_path_recursive): Env Var - depth=0

DEBUG(:enhance_path_recursive): Env Var - max_depth=1

DEBUG(:enhance_path): Env Var INSIDE PATH C:\\<redacted-path>\\Cmder - found=0

DEBUG(:enhance_path): Env Var END PATH C:\\<redacted-path>\\Cmder - found=0

DEBUG(:enhance_path): Appending 'C:\<redacted-path>\Cmder'

DEBUG(:enhance_path): END Env Var - PATH=C:\<redacted-path>\Cmder\vendor\conemu-maximus5\ConEmu\Scripts;C:\<redacted-path>\Cmder\vendor\conemu-maximus5;C:\<redacted-path>\Cmder\vendor\conemu-maximus5\ConEmu;C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\wbin;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Windows\System32\OpenSSH\;C:\Program Files\Intel\WiFi\bin\;C:\Program Files\Common Files\Intel\WirelessCommon\;C:\Program Files\dotnet\;C:\Program Files\Docker\Docker\resources\bin;;C:\<redacted-path>\git-2.39.0\bin;C:\<redacted-path>\git-2.39.0\cmd;C:\<redacted-path>\git-2.39.0\mingw64\bin;C:\<redacted-path>\git-2.39.0\usr\bin;C:\<redacted-path>\Gradle\gradle-7.6\bin;C:\<redacted-path>\Haskell\ghc-8.10.4\bin;C:\<redacted-path>\Java\jdk-17.0.2\bin;C:\<redacted-path>\JetBrains\scripts;C:\<redacted-path>\Kubernetes\bin;C:\<redacted-path>\Make\bin;C:\<redacted-path>\Maven\apache-maven-3.6.3\bin;C:\<redacted-path>\Node\node-18.13.0;C:\<redacted-path>\PHP\php-8.1.3;C:\<redacted-path>\PHP\composer;C:\<redacted-path>\Python\python-3.8.1;C:\<redacted-path>\Terraform;;C:\Users\<redacted>\AppData\Local\Microsoft\WindowsApps;C:\Users\<redacted>\.dotnet\tools;C:\<redacted-path>\Cmder\vendor\bin;C:\<redacted-path>\Cmder

DEBUG(:enhance_path): Env Var C:\\<redacted-path>\\Cmder - found=1

DEBUG(init.bat): Env Var - HOME=C:\Users\<redacted>

DEBUG(init.bat): Calling - C:\<redacted-path>\Cmder\config\user_profile.cmd

Elapsed Time: 0:0:45.32 (45.32s total)
```

**Comment by andrej-dyck** [comment]
Note that this behavior I had first experiened on Windows 10 22H1 when I updated Cmder from an older version. I don't know which version I had before, but with that, the startup-time was 1-2sec.

**Comment by d4k0** [comment]
Same problem here when I started Cmder today after I haven't used it for 2-3 weeks:

![2023-01-22_19-49-34](https://user-images.githubusercontent.com/10982825/213934234-5510267f-2ce8-47ba-920c-346ba56a7b29.png)

Output of `set PATH`:

```
Path=E:\Tools\Cmder\vendor\git-for-windows\cmd;E:\Tools\Cmder\vendor\conemu-maximus5\ConEmu\Scripts;E:\Tools\Cmder\vendor\conemu-maximus5;E:\Tools\Cmder\vendor\conemu-maximus5\ConEmu;C:\Program Files\Java\jdk1.8.0_361\bin;C:\Program Files\Common Files\Oracle\Java\javapath;C:\Program Files (x86)\Common Files\Oracle\Java\javapath;E:\Tools\Alacritty;E:\Tools\apache-ant-1.10.5\bin;C:\Users\*MyUsername*\AppData\Local\bin\NASM;D:\Program Files\CMake\bin;E:\Programme\Video\Konvertieren\FFmpeg;E:\Programme\Dokumente\AtomPortable\Tools\Node.js;E:\Tools\Cmder\vendor\git-for-windows\usr\bin;E:\Tools\Cmder\vendor\git-for-windows\bin;E:\Tools\Cmder;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;C:\Program Files\Microsoft SQL Server\130\Tools\Binn\;C:\WINDOWS\System32\OpenSSH\;C:\Program Files\Crucial\Crucial Storage Executive;C:\Program Files\dotnet\;C:\Program Files\gs\gs9.56.1\bin;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;C:\WINDOWS\System32\OpenSSH\;C:\Program Files (x86)\NVIDIA Corporation\PhysX\Common;C:\Users\*MyUsername*\.cargo\bin;E:\Program Files\Python\Python37\Scripts\;E:\Program Files\Python\Python37\;C:\Users\*MyUsername*\AppData\Local\Microsoft\WindowsApps;E:\Tools\Cmder\vendor\git-for-windows\mingw64\bin;E:\Tools\Cmder\vendor\bin
PATHEXT=.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC
path_position=append
```

**Comment by chrisant996** [comment]
@d4k0 While this probably isn't responsible for the delay, that PATH variable has several duplicates:
- 3 copies of c:\windows 
- 3 copies of c:\windows\system32
- 2 copies of c:\windows\system32\openssh
- 3 copies of c:\windows\system32\wbem
- 3 copies of c:\windows\system32\windowspowershell\v1.0

Repeated directories can slow down launching programs, by searching the directories multiple times.  Cleaning up the variable might be generally useful, but the duplicate directories probably aren't causing 40 seconds of slowdown, so the cause is probably something else yet to be discovered.

In the meantime, if you clean that up, make a backup copy of the contents of the PATH variable first, so that it can be restored if something goes wrong while cleaning it up.

**Comment by chrisant996** [comment]
Can someone try downgrading to an older version of Cmder or ConEmu, and see if the problem goes away?

If downgrading ConEmu resolves the problem, then the issue is likely something in ConEmu, not Cmder.

If downgrading ConEmu doesn't resolve the problem, but downgrading Cmder resolves the problem, then there would be a possibility that the issue might be due to some recent change in the Cmder scripts.

**Comment by andrej-dyck** [comment]
I'll try it tomorrow. 👍

**Comment by chrisant996** [comment]
@DRSDavidSoft @daxgames cmder init debug logging can show elapsed time of the whole init.  Is there a way to see times along the way, to help identify where delays occur?

**Comment by daxgames** [comment]
@andrej-dyck is there really something in your path worth redacting?

**Comment by andrej-dyck** [comment]
> @andrej-dyck is there really something in your path worth redacting? 

Unfortunately, yes. But all paths only use the English letters and some use a space in-between words.

**Comment by daxgames** [comment]
@chrisant996 not currently

**Comment by Hlsgs** [comment]
@chrisant996 I have an older version of mini that i haven't updated in a while and that used to work fine(1.3.16.1035 with conemu 210912). Problem manifests same as with a freshly downloaded version.

@daxgames I've cleaned my path and that didn't help, so that's not it. I'm also unable to replicate this in a fresh VM with an up to date W10. Tricky. I was going to ask the same as @chrisant996 about a timestamped log file...

**Comment by 296951** [comment]
Looking at what it would take to add timestamps to the log

**Comment by 296951** [comment]
@Hlsgs Slowness is USUALLY something on the host that is inspecting every API call, usually some Antivirus or Data Loss Prevention security tool.  We see this a lot on corporate managed systems.

**Comment by 296951** [comment]
Has anyone tried using `/f` on the Conemu task?

**Comment by 296951** [comment]
Just added this to the main branch:

https://github.com/cmderdev/cmder/blob/6bd2e260f0892fcc895c5f67ba949e8597260823/vendor/lib/lib_console.cmd#L43

You can edit your local files and the a timestamped debug output.

**Comment by d4k0** [comment]
> @d4k0 While this probably isn't responsible for the delay, that PATH variable has several duplicates:
> 
>    [...]

Thanks for the hint, I used "Rapid Environment Editor" to clean up the PATH variable (it has a function for this), but the loading time hasn't changed.

I also tried a fresh copy of Cmder mini, but it also loads very slowly (around 25s).

As @mpitz pointed out, Cmder is continuously switching between cmd.exe and findstr.exe.

The only thing that I installed lately was the latest Windows 10 22H2 update last week so this could maybe be the culprit. I will test Cmder on my work notebook tomorrow to see if I have this issue there as well.

**Comment by daxgames** [comment]
If you see `cmder` constantly doing `findstr` AND IT IS SLOW then put `/f` in your conemu task it'll skip trying to prevent duplication in the path and just modify the path.

**Comment by mpitz** [comment]
@daxgames can you point out where to set the parameter exactly?

I've tryied to set it unter SETTINGS>STARTUP>TASKS>Task Parameters. But it says "unsupportet task parameters Task name : {cmd::Cmder} Parameters: /f

**Comment by andrej-dyck** [comment]
Do I undestand this correctly that `/f` doesn't search for `git` in PATH and uses the vendored _'git-for-windows'_? If so, this means that the _'mini'_ version wouldn't have any `git` at all? Shouldn't it be the other way round (I didn't see another option) that says "don't do anything with my `PATH`"?

**Comment by andrej-dyck** [comment]
I have a temporary solution for me. 

I dug deep into the `init.bat`. I'm pretty sure I noticed some bugs and there's a lot of unnecessary work. In essence, I'm skipping almost everything and can reduce the start-up time from ~50sec to 4.5sec.
![image](https://user-images.githubusercontent.com/43913051/214044604-1ee04efe-e8bb-4b1c-90a6-af790b473fbd.png)

Pre-requisites:
* I'm currently on a Windows 10 22H2 machine
* I have git installed and it's in path (%PATH_GIT%\bin, %PATH_GIT%\cmd, %PATH_GIT%\mingw64\bin, %PATH_GIT%\usr\bin)
* I'm using the _'mini'_ version of Cmder
* Make a backup ;)

Here are the changes (skipping mostly Git-related scripts):
* Add `set CMDER_CONFIGURED=2` to the top of the script (this jumps after some initial stuff directly to `:USER_ALIASES`)
![image](https://user-images.githubusercontent.com/43913051/214045177-74b22bdc-7411-4512-9011-57b6d9865489.png)
* **Move** `:PATH_ENHANCE` (l. 343) before first `if "%CMDER_CONFIGURED%" GTR "1"`
![image](https://user-images.githubusercontent.com/43913051/214045487-cb816692-b0e6-4180-af78-5a583d214e4b.png)
* **Move** `if "%CMDER_CONFIGURED%" gtr "1" goto :CMDER_CONFIGURED` (l. 408) to start of `:USER_CONFIG_START`
![image](https://user-images.githubusercontent.com/43913051/214046010-a8f98a09-4809-4dc0-9e65-a460afa45c61.png)

Optional changes:
* Skip `%lib_base% update_legacy_aliases` (l. 383) -- _I have a clean install_
![image](https://user-images.githubusercontent.com/43913051/214046869-b965fad4-380d-4f79-80e5-88208e4f1219.png)
* Skip `echo Migrating alias 'history' ...` (l. 399 - 404) -- _I have a clean install_
![image](https://user-images.githubusercontent.com/43913051/214047027-fda3d494-3eca-4ac7-bd78-b99f6e6c36c3.png)
* Skip `if exist "%GIT_INSTALL_ROOT%\post-install.bat"` (l. 413 - 418) -- _I use the 'mini' version and I'm pretty sure the placement is wrong here as it's only needed for vendored git_

_Note_: referenced line numbers are those in the original file

**Comment by daxgames** [comment]
@mpitz See below:

![image](https://user-images.githubusercontent.com/7318053/214054066-cec2b79c-0741-475b-ba40-ae06fd012330.png)

`/f` reduces `enhance_path` to `set path=...` in that it does not do a `findstr` on the existing path to try and prevent adding things to the path that are already there.

@andrej-dyck  `/f` does nothing to git detection.  Mini will always have git if it is in your path but I think it still tries to do git detection which is unnecessary, If true I will fix,

To avoid Git version detection, and the slowdown that comes with it, on Mini or Full set `GIT_INSTALL_ROOT=[Git install path]` env variable via Windows.

Be careful editing `init.bat` it is overwritten on every update of `cmder`. 

Setting `CMDER_CONFIGURED=2` skips:

- Git version detection and configuration.
  - This is slow
  - Avoidable by setting `GIT_INSTALL_ROOT=[Git install path]` via Windows.
- Add "%CMDER_ROOT%\vendor\bin" to the path configuration.
  - Not super slow by itself, not as fast as it could be.
  - Can be made faster by adding `/f` to the the Cmder task.
- Add "%CMDER_ROOT%\bin" to the path configuration.
  - Not super slow by itself, not as fast as it could be.
  - Can be made faster by adding `/f` to the the Cmder task.
- Running of user `"%CMDER_ROOT%\config\profile.d"` or `"%CMDER_USER_CONFIG%\profile.d"` scripts.
  - If no scripts no delays.
  - Delays are entirely user responsibility.

I would rather add an option for advanced users that would allow them to skip the included `init.bat` entirely and run their own in `%cmder_root%\config\init.bat` if the file exists.  This would give users that want SUPER FAST and COMPLETELY customizable init while allowing Cmder still support the masses and be flexible.

I will also add an additional condition that only runs the `post-install.bat` if using vendored git.  Not sure what good this will do because it should not exist in a pre-installed Git since it is run by the Git installer and deletes itself.

**Comment by andrej-dyck** [comment]
@daxgames thanks for the answer.

> /f does nothing to git detection
In the 'full' version it does: 
![image](https://user-images.githubusercontent.com/43913051/214068300-6500db22-c609-4462-b531-df3e88542b70.png)

> Mini will always have git if it is in your path but I think it still tries to do git detection which is unnecessary, If true I will fix,
Yes, it still tries to do some "vendored" Git stuff (:read_version, :validate_version, :compare_git_versions). So unnecassary in 'mini' and if you have Git in `Path`

> Git version detection [skip] ... Avoidable by setting GIT_INSTALL_ROOT=[Git install path] via Windows.
Well, this reduced my init time from ~27sec to ~23sec. So there's still a lot that's worked on.

> I would rather add an option for advanced users that would allow them to skip the included init.bat entirely and run their own [..]
Well, we could do it right now, can't we?
![image](https://user-images.githubusercontent.com/43913051/214073145-5a7bc883-66f2-4040-8d19-ad94131b1aef.png)

**Comment by andrej-dyck** [comment]
@daxgames I noticed that within ConEmu / Cmder certain commands take a long time (~1sec compared to <50ms natively): 
![image](https://user-images.githubusercontent.com/43913051/214077417-50563126-1ca4-44a8-9b74-e01f3d13be49.png)

That doesn't explain all the unnecessary work in `init.bat`, though.

**Comment by chrisant996** [comment]
Does the slowness happen with any Cmder versions prior to the current version?

If yes, then the cause is not really Cmder, it is something external such as an OS issue or antivirus issue or company-management software.

If stripping all scripts gets down to 4.5 seconds, that is still a very slow number, and should be more like 0.5 seconds.

So, it's starting to look like Cmder is just a victim of something external.

Detailed analysis and troubleshooting would be needed to performed by someone who is able to reproduce the issue.

I'm using Win10 22H2 with no change in Cmder startup, so I cannot repro and am unable to investigate at all.

**Comment by daxgames** [comment]
@andrej-dyck I realized the error in my statement regarding `/f` doing nothing for git detection and edited the original message shortly after I posted it.

> We could do it right now, can't we?

I was thinking something a little more automatic.  As in `%cmder_root%\vendor\init.bat` recognizes that `%cmder_root%\config\cmder_init.bat` exists and runs it else coutinue with the default `%cmder_root%\vendor\init.bat`

@chrisant996 I have been saying it is external all along.  My Cmder startup speed varies wildly but is generally <2-3 seconds even with git version detection enabled.  I have seen it as high as 14 Seconds after launching cmder.exe but the next launch could be 1-2 seconds.

I cannot reproduce the issue so I cannot offer a complete fix.  There are some things that could be eliminated but these will not eliminate the delays reported here.

**Comment by chrisant996** [comment]
> @chrisant996 I have been saying it is external all along.

Yup.  And I'm pretty sure it's external, as well.

The reason I continue to request/suggest additional info from users who are able to reproduce the issue is because being able to successfully demonstrate that it's external can help with redirecting the problem reports to an appropriate place that can do something about them.  And in the unlikely event that additional info pinpoints that it is not an external issue, the addl info would help towards tracking down possible causes (e.g. maybe a batch script change with a non-obvious side effect that only malfunctions on machines with some certain system config).

**Comment by daxgames** [comment]
@andrej-dyck See below after some modifications using Cmder Mini and User installed git with `/f`:

![image](https://user-images.githubusercontent.com/7318053/214203312-f974d97a-91e2-43a4-b05c-fe4d32961424.png)

**Comment by daxgames** [comment]
> @daxgames I noticed that within ConEmu / Cmder certain commands take a long time (~1sec compared to <50ms natively)

We cannot control the speed at which commands run in Conemu.

**Comment by daxgames** [comment]
Unreleased build for #2819 

https://ci.appveyor.com/project/cmderdev/cmder/build/artifacts

**Comment by chrisant996** [comment]
Oh -- I might know what's happening ...

@andrej-dyck what does `clink autorun show` report?

If you installed Clink separately from Cmder and installed it for autorun, then every single launch of cmd.exe will also run `clink inject`.  The first thing `clink inject` does is check if the target cmd.exe process is a non-interactive process, and if so then the inject is cancelled.  But it still takes some time to do that.

It's one of the reasons I recommend to not set up Clink for autorun.  There's nothing Clink can do to improve upon the autorun stuff -- that's completely up to cmd.exe.  AutoRun is convenient, so that you don't have to make shortcuts to start cmd.exe with Clink, but unfortunately AutoRun can significantly slow down cmd.exe invocations.  (If one isn't careful with what's put in the AutoRun regkey, it's even possible to create an infinite loop that consumes all CPU and memory launching thousands of cmd.exe processes reentrantly in a tight loop until the machine reboots -- and again, that's just how cmd.exe and AutoRun work, it's not part of Clink.)

On my machine, when Clink is configured for autorun, then Cmder takes 4 seconds to start, instead of 2 seconds.

Maybe some antivirus software is more aggressively interfering with Clink on your machine and making it slow.  Normally it's very fast to do the inject short-circuit check, but Cmder's init.bat script uses many pipe operations, each of which launches a new hidden non-interactive copy of cmd.exe.  (Even without anything in cmd.exe's AutoRun regkey, the pipes slow things down.)

**So:** if Clink is indeed configured for autorun, then try `clink autorun uninstall`, and see what kind of difference that makes on the speed.

But if Clink isn't configured for autorun, then I'm all out of ideas.

**Comment by andrej-dyck** [comment]
@daxgames First, good job with the PR#2819. 

@daxgames @chrisant996 Yes, it's something external.

I don't think it has something to do with cmder or clink (as far as I can see, ConEmu doesn't come with clink out of the box) and I hadn't had clink installed (now, after this slowliness issue, I installed it and replaced cmder). 

So I have Windows 10 22H2 (with 22H1 I experienced the same slowliness), some software (the big ?), and fresh zips of Cmder (mini) and ConEmu in different versions. All of them executed simple commands very slowly (s. https://github.com/cmderdev/cmder/issues/2816#issuecomment-1400530272). Given that Cmder has a large init.bat, it the slow commands accumulated.

My windows is not managed by any company (i.e., it's a personal Notebook). I turned off my antivirus software; still slow. I turned off any docker services; still slow. I have nothing in autostart, and suspicious services running (firefox updater, chrome updater, antivirus service [that I turned off temp.], lenovo platform service, nvidia). 

Before my recent re-install of the complete system, all services and software for 2-3 years now and the Cmder slowliness appeared a couple of weeks ago. So, my best guess is, that it has something to do with Windows Updates?

**Comment by andrej-dyck** [comment]
@daxgames @chrisant996 Some more information.

While I was writing an response, I started up [Windows Sandbox](https://learn.microsoft.com/en-us/windows/security/threat-protection/windows-sandbox/windows-sandbox-overview), extracted cmder_mini, and it starts in 2sec.
![image](https://user-images.githubusercontent.com/43913051/214273513-51586f86-7d5c-443c-bd2d-9ea038a74eca.png)

However, notice the older Windows 10 version in the sandbox!

Furthermore, on my company notebook, which is managed, has an aggressive antivir and most of the same software, Cmder is quite quick (1-2sec). There, I have Windows 11 installed.

**Comment by rustamsh** [comment]
I am having similar issue. cmder started loading slow without any reason (I didn't install any new updates or something similar). I noticed that cmd.exe, find.exe, findstr.exe, git.exe (and also cmd.exe and findstr.exe restarts about 5-10 times until it is loaded) are loading like 3-4 seconds each and in total it becomes like half a minute or something

**Comment by chrisant996** [comment]
> I am having similar issue. cmder started loading slow without any reason (I didn't install any new updates or something similar). I noticed that cmd.exe, find.exe, findstr.exe, git.exe (and also cmd.exe and findstr.exe restarts about 5-10 times until it is loaded) are loading like 3-4 seconds each and in total it becomes like half a minute or something

@rustamsh If you can capture a performance trace using [Windows Performance Recorder](https://learn.microsoft.com/en-us/windows-hardware/test/wpt/windows-performance-recorder-step-by-step-guide), I'd be willing to take a look and see if I can identify anything specific.

**Comment by andrej-dyck** [comment]
@chrisant996 sorry for not replying. Lot's to do right now. I'm moved on to Windows Command Prompt + Clink + some Clink plugins. I saw that WPR comes with tools that are 2GB+. Seems very heavy and I have a fresh install. I'm not sure what information would be recorded and I don't have the time right now to verify it. Sorry again.

**Comment by chrisant996** [comment]
> @chrisant996 sorry for not replying. Lot's to do right now. I'm moved on to Windows Command Prompt + Clink + some Clink plugins. I saw that WPR comes with tools that are 2GB+. Seems very heavy and I have a fresh install. I'm not sure what information would be recorded and I don't have the time right now to verify it. Sorry again.

Understood; no worries.

What it would record is a performance trace of all the APIs that are called by processes and the timings for them, so that they can be viewed/searched/analyzed in a timeline view (and other views).  That would make it possible to identify specifically how the additional time is being spent.

P.S.  I thought the WPR install can be customized to be smaller, but I don't remember for sure.  And like you, I have a lot going on and I'm not going to dig into that, specifically.  🙃

**Comment by andrej-dyck** [comment]
> Understood; no worries.

Thanks.
 
> What it would record is a performance trace of all the APIs that are called by processes and the timings for them, so that they can be viewed/searched/analyzed in a timeline view (and other views).  That would make it possible to identify specifically how the additional time is being spent.

I think I found a cli version with sysinternal tools (not sure) but it was all confusing to me how to start measuring cmder with it. I might try it out again once I find spare time. 

> P.S.  I thought the WPR install can be customized to be smaller, but I don't remember for sure.  And like you, I have a lot going on and I'm not going to dig into that, specifically.  🙃

**Comment by chrisant996** [comment]
> I think I found a cli version with sysinternal tools (not sure) but it was all confusing to me how to start measuring cmder with it. I might try it out again once I find spare time.

Do you mean ProcMon? That's different, and my understanding is its output cannot be consumed by the Windows Performance Analyzer toolset. It might offer some info about what's happening, and might be useful, but it would be more limited, and analyzing the output would be much harder and manual.

**Comment by metinsut** [comment]
same issue,
any improvement?

**Comment by andrej-dyck** [comment]
On windows 11? That's new to me. It seemed to me that it only happened on win10 22h1+2

**Comment by WickedSilver** [comment]
For me it's on Windows 10 PC as well as on my Windows 11 laptop.
Both are work PCs and are registered under a domain, but never really had this amount of slowness until recently.
The only visible difference between my PC and Laptop is the following while starting cmder:
- On my PC it keeps switching from cmd to find 
- On my laptop it switches from cmd to findstr.

I checked out the performance tool and found the following application (in Dutch). Is this the right one? If so can you tell me what i can do to help?
![image](https://user-images.githubusercontent.com/2571093/218088778-0b5bbf99-9243-4cd4-85c1-9f62fb89c21a.png)

**Comment by daxgames** [comment]
@metinsut @WickedSilver have either if you tried `/f` as mentioned above.  It is nit a fix but it helps.  Also as stated above it is a problem with something external to Cmder so not much we can do unless someone with the problem can track down the actual cause.

**Comment by WickedSilver** [comment]
> @metinsut @WickedSilver have either if you tried `/f` as mentioned above. It is nit a fix but it helps. Also as stated above it is a problem with something external to Cmder so not much we can do unless someone with the problem can track down the actual cause.

Thanks for the answer, and i just tried with /f and starting or opening a new tab takes ~15 seconds under Windows 10 / PC.
I'm pretty sure it's ~30+ seconds without but didn't time it.

I can try to help track down the actual cause, but not sure how.

**Comment by daxgames** [comment]
@WickedSilver you can add `/t` to get exact time it takes to start

**Comment by chrisant996** [comment]
> I checked out the performance tool and found the following application (in Dutch). Is this the right one? If so can you tell me what i can do to help? 

That looks like Performance Monitor (perfmon), which is for a very different kind of thing and won't be helpful here.

[Windows Performance Recorder](https://learn.microsoft.com/en-us/windows-hardware/test/wpt/windows-performance-recorder) is what I was referring to.  It's part of a fairly big installation of diagnostic toolkit stuff.

**Comment by daxgames** [comment]
I did some more optimizing tonight #2825.  This build is twice as fast as the current `master` branch for me at loading.

Screenshots are in the PR showing load times.

Please test and provide feedback

Get downloads [here](https://ci.appveyor.com/project/cmderdev/cmder/builds/46204186/artifacts)

**Comment by daxgames** [comment]
@usoliveira I am working on a way that is even faster for users who want to be in complete control and do not need all of Cmder ability to do auto configuration.

It is no good for the first time/inexperienced user but for someone seeing slowness due to some unknown external cause with full auto config it can be useful.

Users can control the config COMPLETELY using an optional hard coded script in `%cmder_root%\config` that still supports everything `%cmder_root%\vendor\init.bat` does.  If present `%cmder_root%\vendor\init.bat` runs the optional script and exits skipping all the auto config that makes Cmder very flexible.

I have seen load times of less than 1 second using the same VM in the PR.

**Comment by Ali-Nategh** [comment]
> I did some more optimizing tonight #2825. This build is twice as fast as the current `master` branch for me at loading.
> 
> Screenshots are in the PR showing load times.
> 
> Please test and provide feedback
> 
> Get downloads [here](https://ci.appveyor.com/project/cmderdev/cmder/builds/46204186/artifacts)


I downloaded and tested this, it is still slower than the version I had started with but it is faster than the updated main version where the bug begun.

the bottom left tab cycling between cmd.exe, find.exe, findstr.exe, git.exe, is still here though.

```
  Elapsed Time: 0:0:16.04 (16.04s total)
  
  Clink v1.4.16 is available.
  - To apply the update, run 'clink update'.
  - To stop checking for updates, run 'clink set clink.autoupdate false'.
  - To view the release notes, visit the Releases page:
    https://github.com/chrisant996/clink/releases
```
after clink update:

```
  Elapsed Time: 0:0:16.03 (16.03s total)
  
  D:\Ali\Downloads\cmder
```

Windows version: 
```
Edition	Windows 11 Pro
Version	22H2
Installed on	‎12/‎29/‎2022
OS build	22621.1105
Experience	Windows Feature Experience Pack 1000.22638.1000.0
```

Thanks for the update, if you need any further information let us know. 🤝

**Comment by Ali-Nategh** [comment]
"/f" reduces it to 10.56s

**Comment by chrisant996** [comment]
@andrej-dyck regarding these:

> My windows is not managed by any company (i.e., it's a personal Notebook). I turned off my antivirus software; still slow.

Did you turn off Windows Defender?  I ask in case "turned off my antivirus software" refers to AV software that you installed separately.

> I have nothing in autostart

Do you mean OS autostart?
Or do you mean CMD AutoRun?
What does `clink autorun show` report?
Has `clink autorun uninstall` been tried?

<br/>
<br/>

Maybe this list of conclusions and questions will help:

### What's Known So Far

- Old versions of Cmder exhibit the problem suddenly, as well.
  - **Conclusion:** This is not due to any recent change in Cmder or ConEmu.
- Multiple builds of Win10 and Win11 exhibit the problem, but only for some people.
  - **Conclusion:** This might be related to some OS change, but it isn't happening universally for everyone with those OS versions.
- The slowness is evident in the invoking of programs such as `findstr` and piping their output.
  - **Conclusion:** The fact that "findstr" takes long enough to show up in the title suggests that the slowness is actually happening to `findstr` and the other invoked programs.

### Pending Questions

- Is there a connection with Clink?
- Is there a connection with Windows Defender?
- Does slowness only occur during Cmder's `init.bat` script, or are similar commands slow after that, as well?
- Does slowness only happen on the first `init.bat` invocation in a session?  What happens if you `set CMDER_CONFIGURED=` and then run `vendor\init.bat` again?
- Does slowness happen when invoking Cmder's `init.bat` script from a plain old `cmd.exe` without using ConEmu?

**Comment by daxgames** [comment]
More improvements in #2825 

- Added a new lean yet full featured optional `%cmder_root%\vendor\user_init.optional.cmd`
  - Usage documentation is in the file.
  - Basically an almost completely hard coded version of the `%cmder_root%\vendor\init.bat` that gives the user complete control if they want it.
  - See speed improvements from `main` -> `more_speed_2` -> `more_speed_2 with user_init.cmd` below with `/t` only:
    ![image](https://user-images.githubusercontent.com/7318053/218290724-0085e15d-cd6e-44eb-b442-be747053cffa.png)

**Comment by andrej-dyck** [comment]
@chrisant996 

> Did you turn off Windows Defender? I ask in case "turned off my antivirus software" refers to AV software that you installed separately.

Both. I turned of my AV software and Windows Defender.

> Do you mean OS autostart? Or do you mean CMD AutoRun?

OS Autostart.

> What does clink autorun show report? Has clink autorun uninstall been tried?

Autorun wasn't / isn't active as clink wasn't installed (only what has been shipped with the cmder zip).

> Is there a connection with Clink?

I don't think so, because I have now Windows Command Prompt + Clink installed and it works very fast. 

> Does slowness only occur during Cmder's init.bat script, or are similar commands slow after that, as well?

No, all the commands are executed slower than in native Command Prompt / Powershell. It's only because init.bat used a lot of commands, the time accumilated and was more noticable. See my comment https://github.com/cmderdev/cmder/issues/2816#issuecomment-1400530272. Since all commands are executed more slowly, also your next two questions are answered.

> Does slowness happen when invoking Cmder's init.bat script from a plain old cmd.exe without using ConEmu?

I'm not sure. I configured `"cmd.exe" /k ""%CMDER_ROOT%\vendor\init.bat""` in my IDE to use my Cmder aliases. And it was slow. But again: I'm not sure if this answers your question.

**Comment by chrisant996** [comment]
> > Does slowness happen when invoking Cmder's init.bat script from a plain old cmd.exe without using ConEmu?
> 
> I'm not sure. I configured `"cmd.exe" /k ""%CMDER_ROOT%\vendor\init.bat""` in my IDE to use my Cmder aliases. And it was slow. But again: I'm not sure if this answers your question.

Yup, that answers the question, thanks!

At this point, definitely the problem is not related to Cmder, nor to ConEmu, nor to Clink.

What a frustrating situation! 🫤

**Comment by andrej-dyck** [comment]
> > > Does slowness happen when invoking Cmder's init.bat script from a plain old cmd.exe without using ConEmu?
> > 
> > I'm not sure. I configured `"cmd.exe" /k ""%CMDER_ROOT%\vendor\init.bat""` in my IDE to use my Cmder aliases. And it was slow. But again: I'm not sure if this answers your question.
> 
> Yup, that answers the question, thanks!
> 
> At this point, definitely the problem is not related to Cmder, nor to ConEmu, nor to Clink.
> 
> What a frustrating situation! 🫤

That's strange. Note that I noticed a slow start in the IDE terminal but didn't measure the execution time of single commands.

Here's why I find it strange. On the same machine, with the same git installation and its MinGW-w64, the git bash, CMD+clink, and PowerShell works very fast.  I also tried other shell emulators for windows (which I didn't like) and they were also fast. Only ConEmu and Cmder executed the commands slow.

Again, since I moved on 🥲, I can't do anymore experiments...

**Comment by chrisant996** [comment]
> Here's why I find it strange. On the same machine, with the same git installation and its MinGW-w64, the git bash, CMD, CMD with Clink, and PowerShell work very fast.  I also tried other shell emulators for windows (which I didn't like) and they were also fast. Only ConEmu and Cmder executed the commands slow.

To be clear, ConEmu and Cmder aren't executing the other programs.  They ask the OS to launch them, and the OS executes them.

Since this suddenly started having even to old versions of Cmder and ConEmu that had been working fine, that shows that something external has changed and is causing the slowness.

Also, Cmder is essentially a configuration package, with an initialization script.  Since that same initialization script (`vendor/init.bat`) runs at normal speed outside ConEmu, that shows the issue is unrelated to Cmder.

The fact that the slowness is only in the context of ConEmu is interesting.  But ConEmu isn't executing the programs, it only displays their output on the screen.

However, ConEmu does have settings that control whether it injects its DLLs into console programs.  Maybe something is interfering with that.

Have you tried turning those off?
- [Settings > Integration > Default term](https://conemu.github.io/en/SettingsDefTerm.html)
  - Force ConEmu as default terminal for console programs
  - Leave in TSA
  - Aggressive mode
  - List of hooked executables...
  - Don't use ConEmuHk.dll in started console

**Comment by daxgames** [comment]
More updates.  I am very excited about this!

Checkout #2825 Init.bat and all its bloat is now a run-once thing!

**Comment by chrisant996** [comment]
> More updates. I am very excited about this!
> 
> Checkout #2825 Init.bat and all its bloat is now a run-once thing!

Right on!  Nice work!!  🥳🤩

**Comment by mtaki14** [comment]
I'm experiencing the issue as well (win11). The latest changes daxgames made have definitely improved things a lot but it still takes 5+ seconds to initialise cmder for me. The improvements are a nice byproduct of this issue but I'm still struggling to find the actual cause.

Runnint init.bat in cmder takes 4.25s meanwhile running it in powershell/cmd takes 0.25s

> Have you tried turning those off?

> [Settings > Integration > Default term](https://conemu.github.io/en/SettingsDefTerm.html)
> Force ConEmu as default terminal for console programs
> Leave in TSA
> Aggressive mode
> List of hooked executables...
> Don't use ConEmuHk.dll in started console

I tried playing around with these but it made no difference unfortunately

**Comment by daxgames** [comment]
@mtaki14 Just curious what was the time before the latest change?

Unfortunately there is nothing else I can do to make it faster as I am not really experiencing what you are seeing.  

The `init.bat` was trying to be all things to everyone and had some inherent slowness due to the auto detection it was doing as evidenced by the latest build.

The file that was generated is yours to edit.  You can add/remove whatever you want, but the cmd/powershell times you posted tell me you are not going to make it faster.

@chrisant996 has offered to look at a trace of the processes if someone experiencing the problem can take one.  Please look at messages above if you are willing to try and provide one.

We want to help solve this, we have done all we can at this point since none of us are experiencing the core issue so we need help from someone experiencing the problem.

**Comment by daxgames** [comment]
Times on my SUPERFAST AMD Ryzen 9 5950x with Samsung 980 Pro SSD Desktop:

![image](https://user-images.githubusercontent.com/7318053/219043008-cc09af6b-8cf4-4d1d-bf5d-cf94d0f0633c.png)

**Comment by daxgames** [comment]
The same tests as above on my dev vm:

![image](https://user-images.githubusercontent.com/7318053/219050555-e537bc71-f00b-4358-9756-3412ad2173f9.png)

**Comment by daxgames** [comment]
@usoliveira what kind of improvements are you getting?

**Comment by daxgames** [comment]
@usoliveira thats still really slow but better than 13 seconds for sure.

**Comment by Crayder** [comment]
Same things here. Exactly as people above described. Starting about start of last month it takes 50 seconds to launch or so. Only on the init.bat startup. Loops between `cmd.exe` and `findstr.exe` as someone above said.

![image](https://user-images.githubusercontent.com/9919010/219871837-a1b6fda3-81c8-4f55-8463-b8612f2744bc.png)

It's not a coincidence we all had this problem come up at the same time, right? There has to be a difference somewhere that we're missing.

**Comment by daxgames** [comment]
@Crayder As stated above something changed in Windows builds that caused a less than optimal `%cmder_root%\vendor\init.bat` to become ridiculously slow for a subset of users.  We have been unable to determine the root cause.

My code is not a fix it is a fundamental change in the way Cmder initializes `cmd.exe` sessions.

It is for ALL Cmder users regardless of how long it takes to initialize!  All users will see an improvement in first launch speeds and second launch speeds are even faster as shown by the screenshots above.

Latest build [here](https://ci.appveyor.com/project/cmderdev/cmder/builds/46271978/artifacts)

**Comment by chrisant996** [comment]
> It's not a coincidence we all had this problem come up at the same time, right? There has to be a difference somewhere that we're missing.

Correct. But the change was external. Something is interfering with programs that are launched from within ConEmu. But it suddenly started happening to all versions of ConEmu, even old versions, so it's an external change. Cmder team cannot fix the external problem.

> EDIT: After further reading through the above just wanted to note. My startup times were NEVER this slow until the past ~month. On two Win10 machines, on two different networks. Right after I convinced my coworker to start using Cmder might I mention, unfortunate timing for sure. daxgames' fix above is for 5-10s slowness from my understanding, what about us 50 second folk?

Daxgames' fix removes most of the startup work, so that it only happens on the first startup. It benefits everyone. It might reduce your 50 second startup significantly. The only way to know how much it will help your specific case is to try it.

**Comment by XpecV52** [comment]
Maybe this has something to do with the windows system update. I am using the windows 10 professional workstation version on my personal laptop (automatic updates have been disabled for a long time), and the loading speed of cmder is as fast as before.
But the office computer is also a windows 10 system and it becomes very slow until the loading is complete.
After I cleared and reset all the 'Startup->Tasks' in the settings, the console that comes with the windows system will be selected by default at startup, which improves the loading speed.
![image](https://user-images.githubusercontent.com/43233434/220806799-89ec1cae-208a-4fbc-b9d4-79e11491eaf5.png)

**Comment by daxgames** [comment]
@XpecV52 You have essentially removed `Cmder` from your `Conemu` config.  This thread is no longer relevant with your displayed config.

**Comment by deniscuculic** [comment]
Too bad you are not able to find a workaround.
Anyone can recommend any alternatives to Cmder?

**Comment by daxgames** [comment]
@deniscuculic, as stated above several times, this is not a Cmder problem.

That being said, this thread contains links to an unreleased build that removes some inefficiency in the Cmder init process.  I am seeing load times of .19  seconds.  That is pretty fast.  Prior to this build U routinely saw launch times of 1.7-2.5 seconds.

The problem is that none of the Cmder developers can reproduce the problem to troubleshoot it adequately.

**Comment by deniscuculic** [comment]
Than it seems Cmder is dead, and it is time to switch to something else.

I tested Cmder or several windows PCs, with latest updates, and it is same slow behaviour on all PCs.

Other terminal emulators are not affected as Cmder, so I would say it is not only external problem.

**Comment by chrisant996** [comment]
> Than it seems Cmder is dead, and it is time to switch to something else.
> 
> I tested Cmder or several windows PCs, with latest updates, and it is same slow behaviour on all PCs.
> 
> Other terminal emulators are not affected as Cmder, so I would say it is not only external problem.

To clarify -- this affects programs launched from inside _ConEmu_.  It happens regardless whether the Cmder extensions to ConEmu are used.  The Cmder repo team isn't the team who could really investigate or do anything here -- the ConEmu repo is where this needs to be reported.

**Comment by daxgames** [comment]
@deniscuculic,

1.  Cmder is not a terminal emulator.
2.  It is a set of init scripts developed by the Cmder team that ties together 4 externally developed components.
   - A terminal emulator called Conemu.
   - Shell enhancements clink and clink completions.
   - Git for Windows in the full version.
   - Packaged together as a portable shell env called Cmder.
3.  The EXTREME slowness reported in this thread started happening AFTER a specific Windows update to current and all previous Cmder releases.  This indicates it is indeed an external issue.

I can run the Cmder init scripts in native Cmd.exe, VSCode Terminal, Tabby, Conemu, Windows Terminal, and Hyper.  

All have EXACTLY THE SAME speed results I mentioned above with full Cmder functionality.  Even on an old underpowered VM on a laptop with a slow hdd.

We are here to help, but we have done EVERYTHING we can since none of us are experiencing the problem and never have.

Instead of being so negative and trying to piss someone off, how about you tell us how it is performing for you?

Are you seeing extreme slowness reported above?  Please provide launch times by adding `/t` to the task used to start Cmder as it provides the most accurate launch times.

If so, can you provide the traces asked for above by @chrisant996 that might point to an issue we can not see?

**Comment by d4k0** [comment]
@daxgames 
I tried to recreate the problem in a VM with Windows 10 22H2 and all updates. However, to my surprise, Cmder started very quickly there.

Therefore, I have now installed Windows Performance Analyzer on my production system where the problem occurs and created a measurement.

However, I can't really detect any real "bottleneck". It is noticeable that apparently many instances of cmd.exe and findstr.exe are started. Here is a screenshot:

![wpa_2023-03-12_21-42-19](https://user-images.githubusercontent.com/10982825/224576113-a5915974-8fe4-408e-937b-dbfbc5cb584c.png)

I could also provide the ETL file.

**Comment by chrisant996** [comment]
> Therefore, I have now installed Windows Performance Analyzer on my production system where the problem occurs and created a measurement.
> 
> However, I can't really detect any real "bottleneck". It is noticeable that apparently many instances of cmd.exe and findstr.exe are started. 
>
> ...
> 
> I could also provide the ETL file.

@d4k0 can you either post the ETL here or email it to the address in my GitHub profile? I'll take a look and share whatever I'm able to observe!

**Comment by d4k0** [comment]
@chrisant996 
I sent you a link to the file via email. Maybe you can isolate the problem or find the culprit.

**Comment by daxgames** [comment]
@d4k0 have you tried the latest [build](https://ci.appveyor.com/project/cmderdev/cmder/build/artifacts) linked in this issue.

**Comment by chrisant996** [comment]
Here are some questions / ideas / observations based on reviewing the WPA ETL I received from @dk40.

Later tonight, I'll capture a similar trace on my machine where processes are fast within ConEmu, and compare the performance traces to look for similarities and differences.  Hopefully that will reveal some things to explore further.

**Does the speed improve if you temporarily turn off ESET?**
- The files `ebehmoni.dll` and `ebehmonl.dll` are loaded into each process.
- The files are part of the ESET anti-malware suite.
- `ekrn.exe` (the ESET Kernel) is the largest consumer of CPU time, and it correlates with each process start.  That's to be expected, but I can't tell whether it's interfering with the speed of execution.
- It's also possible that ESET is suspicious about ConEmu injecting `conemuhk.dll` into each process.  Turning off ConEmuHk injection might also be able to improve the speed, especially if the slowness is coming from ESET.

**lsass.exe network operations**
- There appears to be a steady stream of roughly 1 second long network Send requests from `lsass.exe` during the trace.  They roughly align with just after each of the `cmd.exe` and `findstr.exe` etc processes starts.  It almost looks as though some kind of authentication is happening that is correlated with starting processes.
- This looks surprising to me.  There is correlation, but [correlation does not imply causation](https://en.wikipedia.org/wiki/Correlation_does_not_imply_causation).

**Does the speed improve if Cmder is installed in `C:\Cmder\`?**
- Cmder was installed into `%USERPROFILE%\Downloads\cmder_mini\`.  The search indexer (and backup programs and other programs) watch for activity on files in the user's Downloads and Documents directories and other user profile directories, and they may trigger additional work.
- I see no speed difference when running Cmder from `d:\tmp\Cmder\` or `%USERPROFILE%\Downloads\Cmder\`, but I don't have ESET installed.

**Something looks potentially surprising with SearchIndexer and/or Prefetch**
- SearchIndexer.exe appears to be getting woken repeatedly due to changes in Prefetch files `cmd.exe-4a81b364.pf` and `findstr.exe-2e9c6fe2.pf` files, and a few other prefetch files.
- I don't understand why SearchIndexer is monitoring for file change notifications in `C:\Windows\Prefetch\`, but maybe that's expected and normal?

**Other observations**
- SmartScreen.exe starts right before ConEmu starts.  SmartScreen might be doing extra verifications, and might also do extra verifications on all programs started from within ConEmu.  **Have you verified for each individual .exe and .dll file in all subdirectories under `cmder_mini`, that each of them has been successfully Unblocked?**  I.e. that none of them show an "Unblock" button in their individual File Properties dialogs.
- PunkBuster (`PnkBstrA.exe`) appears to be doing a heartbeat of network IO, roughly once per second.
- GamingServices.exe seems to be doing some kind of work in the background as well, including some network IO.

**Comment by mtaki14** [comment]
> Does the speed improve if you temporarily turn off ESET? 

Pausing ESET protection temporarily didn't make a perceivable difference for me

**Comment by chrisant996** [comment]
Follow-up from a WPA trace on my machine that does not encounter the slowness:
- There's a lot of lsass.exe network activity in my trace, but it's all very fast (longest was about 70 milliseconds) -- nothing anywhere near the 1000 millisecond durations in dk40's WPA trace.
- SearchIndexer gets triggered similarly; .pf Prefetch files are triggering data change notifications and waking up SearchIndexer.  So that seems to be present regardless whether it's operating slowly.

**If you disconnect ethernet and put the computer into Airplane Mode, does that improve the speed problem?**

**Comment by mtaki14** [comment]
> If you disconnect ethernet and put the computer into Airplane Mode, does that improve the speed problem?

No change. With Eset disabled, Ethernet disconnected and computer in Flight Mode, cmder start-up time is still 5s+, same as before

**Comment by WickedSilver** [comment]
I can confirm that disabling HIPS (Host-based Intrusion Prevention System) in ESET fixed the slowness for me.
Disabling HIPS does require a restart though. It might be best to add a rule to exclude cmder/conemu from ESET instead of disabling it altogether, but not sure how to do that at the moment.

Thanks for the reported observations @chrisant996!

**Comment by mtaki14** [comment]
Wow @WickedSilver is right, disabling HIPS fixed the issue for me as well. I assumed that pausing protection stopped everything ESET related, I guess not then.

I am too currently trying to figure out how to whitelist it. Adding the whole cmder folder as a hips exclusion doesn't seem to do the trick.

**Comment by mtaki14** [comment]
Still haven't found a way to adequately whitelist it but it seems the issue comes specifically from the option "Deep Behavioral Inspection", which is found under the HIPS section of advanced settings. Disabling that pretty much restores cmder's speed for me.

**Comment by minstyle** [comment]
I have also ESET Endpoint Antivirus installed on my machine!

**Comment by d4k0** [comment]
> @d4k0 have you tried the latest [build](https://ci.appveyor.com/project/cmderdev/cmder/build/artifacts) linked in this issue.

I tried it and it significantly reduces startup time (17s vs 5s). Great work!

> Still haven't found a way to adequately whitelist it but it seems the issue comes specifically from Eset's "Deep Behavioral Inspection", which is found under the HIPS section of advanced settings. Disabling that pretty much restores cmder's speed for me. No restart is required when disabling/enabling that option, like when disabling the whole HIPS engine.

I can confirm that the cause is ESET's "Deep Behavioral Inspection". I installed ESET in a VM and then the startup time of Cmder Mini (latest build) was also suddenly 5s.

After disabling "Deep Behavioral Inspection" the startup time is about 0.7s.

But since this is not a permanent solution for me, **I played around with the exceptions and probably found a solution. It is enough to add the path to cmd.exe (C:\Windows\System32\cmd.exe) under the exceptions for "Deep Behavioral Inspection".**

If I add ALSO the folder of Cmder there, the startup time does not improve further. If I ONLY add the folder of Cmder, the startup time improves only minimally (about 0.5s faster, about 4.5s in total).

The question now is: is the cause due to a Windows update (patch day was January 10) or was it due to the last ESET program update on January 16? I have been using ESET for a year now and never had any problems with it in conjunction with Cmder until mid-January. I'm just a little puzzled that the cmd.exe exception "solves" the problem. Maybe we should inform ESET support?

**Comment by chrisant996** [comment]
@d4k0 ah, a smoking gun 😎 yes a Jan 16 update of ESET fits the timeline with the first report here on Jan 18.

**Yes, I would definitely recommend informing ESET support about this severe performance problem.**

<br/>
<br/>

Also, a sidebar -- it's is an interesting point for troubleshooters to be aware of for future issues:  protection suites may require multiple actions to turn them off to the point where they're doing zero work.  Until they're doing zero work, they're not really "turned off" enough to rule out whether they're contributing to or causing the problem.  Unfortunately, since each suite is different, it might take a bit of Q&A in each incident to identify which suite is in use and how to fully turn it off.

On Jan 24 @andrej-dyck said "I turned off my antivirus software; still slow."  I expect that test probably suffered the same wrinkle that occurred in @mtaki14 's test where pausing protection doesn't fully turn off the ESET suite, and the Host-based Intrusion Prevention System remains on (and also a reboot might be necessarily when turning off ESET for it to fully take effect).

**Comment by deniscuculic** [comment]
This also solves the problem for me :) Now it's an instant startup.
@d4k0 , good catch!!

![eset_problem](https://user-images.githubusercontent.com/16257947/225142155-f2f04811-8959-4ef5-8b1e-dca0060127dd.png)

**Comment by daxgames** [comment]
I guess we need to start asking specifically which antivirus are you running when people report slownes

Seems like a bit of a pattern here to me.

**Comment by iki** [comment]
@daxgames Confirming relation to ESET deep behavioral inspection and fix by `C:\Windows\System32\cmd.exe` exclusion here

Also before the ESET HIPS config workaround, setting `CMDER_USER_FLAGS=2` to skip Git config and nix tools path mangling helped greatly (providing you already have your git/bash path setup).

That said, the issue may actually point to actual bugs in current Cmder scripts.

For example, when I enabled `echo on` in [vendor/bin/cexec.cmd](https://github.com/cmderdev/cmder/blob/master/vendor/bin/cexec.cmd) I saw it's delaying on [line 53](https://github.com/cmderdev/cmder/blob/master/vendor/bin/cexec.cmd#L53) evaluated to
```cmd
echo     | %WINDIR%\System32\find /i " ">nul
```
after calling `cexec.cmd /setPath`, probably because of missing `exit /b 0` after [line 21](https://github.com/cmderdev/cmder/blob/master/vendor/bin/cexec.cmd#L21)?

Another big delay is in `vendor/lib/libbase.cmd` and the git/nix config mentioned above.

One of suspects are the `echo %list% | %WINDIR%\System32\find /i "%arg%">nul` constructs, which can be effectively rewritten like
```cmd
if defined arg for %%a in (%list%) do if /i "%%~a"=="%arg%" ...
```

**Comment by daxgames** [comment]
@iki thanks for the input.  I am not familiar with the code referrences but will have a look.

Have you tried the unreleased build mentioned in the is issue?

**Comment by iki** [comment]
@daxgames no, I still have v1.3.20 from October, the chocolatey package [cmdermini](https://community.chocolatey.org/packages/cmdermini) is not updated on each release. 

Should I try with the [build](https://ci.appveyor.com/project/cmderdev/cmder/builds/46271978/artifacts) you referenced, or with the [latest](https://ci.appveyor.com/project/cmderdev/cmder) one?

**Comment by daxgames** [comment]
@iki the current build of PR #2825 is best.

This makes init.bat a 'run-once' auto-configure so the first run is the slowest but after that, mine loads in .5-.75 seconds.

I am not running `Eset` and before this PR my Cmder load times were 1.5-3.5 seconds depending on what elese was going on on the system.

**Comment by pmsobrado** [comment]
> @iki the current build of PR #2825 is best.
> 
> This makes init.bat a 'run-once' auto-configure so the first run is the slowest but after that, mine loads in .5-.75 seconds.
> 
> I am not running `Eset` and before this PR my Cmder load times were 1.5-3.5 seconds depending on what elese was going on on the system.

Where can I find that build?

**Comment by daxgames** [comment]
@pmsobrado https://ci.appveyor.com/project/cmderdev/cmder/builds/47227774/artifacts

**Comment by pmsobrado** [comment]
Thanks!

**Comment by daxgames** [comment]
I don't think so. That was fixed 4 months ago.

**Comment by daxgames** [comment]
> I don't think so. That was fixed 4 months ago.

**Comment by pmsobrado** [comment]
@daxgames You're right, I don't have that error, but I do have a line every time I run cmder saying "\vendor\bin\create-cmdercfg.ps1" wasn't expected here (or now, can't remember).

**Comment by daxgames** [comment]
@pmsobrado where is your cmder folder?

**Comment by pmsobrado** [comment]
@daxgames My absolute path has an space on it, it that's what you mean. But it has never game me errors before:

C:\Program Files (x86)\Cmder (Portable)

**Comment by pmsobrado** [comment]
@daxgames Sorry about that, thanks!

**Comment by daxgames** [comment]
NP did it help?

**Comment by pmsobrado** [comment]
Yes, but launching it like cmder.exe /C has the effect of missing the lambda λ symbol on the prompt, which I liked :(

**Comment by daxgames** [comment]
Thats odd.  It should not have that effect.

**Comment by chrisant996** [comment]
@pmsobrado Maybe the new config directory hasn't been set up.  Maybe it's a side effect of init.bat doing initialization stuff only the first time.  Maybe you can try copying the contents of the old config directory to the new config directory.

**Comment by pmsobrado** [comment]
I tried a clean install on an user folder. It seems to give the ps1 error when the path have spaces, unrelated to permissions. About the lambda, it works when launching normally but not with the /C command (clean install).

**Comment by daxgames** [comment]
@pmsobrado you found another bug in this branch.  I'll fix that.

**Comment by pmsobrado** [comment]
@daxgames Yes please, because launching Cmder with /C is also the only way for clink to work. I have it on my system and if launching Cmder normally, apart from the ps1 error if on a path with spaces (may also be a bug since it's not related to a write protected folder?), I also have this error:

ERROR: Clink initialization has failed with error code: 1

But not when launching with /C.

Thanks for all your efforts!

**Comment by daxgames** [comment]
Fixed [here](https://ci.appveyor.com/project/cmderdev/cmder/builds/47456517/artifacts) as soon as it's done building:

![image](https://github.com/cmderdev/cmder/assets/7318053/fea28772-8ef2-4c76-a57d-e051be5f334a)

**Comment by pmsobrado** [comment]
Thanks, the lambda bug is resolved but the app still do weird things when in a path with spaces, and also the problem with clink is still there:

ERROR: Clink initialization has failed with error code: 1

**Comment by daxgames** [comment]
@pmsobrado You need to be a LOT more specific because it seems to work perfectly for me and Cmder supports being in a folder path with spaces.

My Cmder is in `C:\Program Files\Cmder` so I have spaces in my path also.

I launched using the following command:

`"C:\Program Files\Cmder\Cmder.exe" /c "%USERPROFILE%\.config\cmder"`

I need descriptions of the issues experienced, step by step replication instructions and screenshots.

**Comment by pmsobrado** [comment]
@daxgames Sure, here it is:

Downloaded cmder_mini.zip, extracted it's contents to "C:\Users\Sobrado\cmder test".

First run (just double clicking on the cmder exe):
![imagen](https://github.com/cmderdev/cmder/assets/17699739/8cc6e13a-97f0-4511-8963-851e1de02321)

Second run:
![imagen](https://github.com/cmderdev/cmder/assets/17699739/376e6697-bdd2-438d-a00f-2839cf7f8c99)

You are absolutely right, no problems with a path with spaces, but, I do have a problem with clink, which is installed in my system apart from cmder. I have the same problem when I start it as in your example.

If clearing this option:
![imagen](https://github.com/cmderdev/cmder/assets/17699739/927f9fee-2cd9-4094-9604-4ed72d55a4a7)

it works fine, but I lose clink in normal cmd windows.

**Comment by daxgames** [comment]
@chrisant996 I don't use clink outside Cmder.  Any insight here?
]

**Comment by pmsobrado** [comment]
@daxgames Maybe I should open a new issue for that? I don't know what more info to give, but I will provide anything you guys ask me.

**Comment by chrisant996** [comment]
@daxgames  That's https://github.com/cmderdev/cmder/issues/2800, and it's a bug in Cmder, not a Clink issue.  I assume @DRSDavidSoft is busy and hasn't gotten around to fixing it.

@pmsobrado  The "Clink initialization has failed" message is a bug in Cmder.  Apart from looking unsightly, everything is actually fine.  The Cmder script misinterprets Clink's response, mistakenly thinks Clinks failed when Clink is configured for autostart, and the Cmder script says a failure occurred, but actually everything is fine.

**Comment by daxgames** [comment]
@chrisant996  so I should jut remove the error checking?

**Comment by pmsobrado** [comment]
@daxgames That would be nice, even if it's only a cosmetic thing, at least for this independent build :P

**Comment by daxgames** [comment]
@chrisant996 @pmsobrado Let me look at the linked issue more closely and figure out what to do.  Not going to be able to immediately look at it though.

**Comment by pmsobrado** [comment]
No rush! Thanks!

**Comment by daxgames** [comment]
Fixed [here](https://ci.appveyor.com/project/cmderdev/cmder/builds/47456909/artifacts) when its finished building.

**Comment by pmsobrado** [comment]
Damn, what took you so long? 😄

**Comment by chrisant996** [comment]
@daxgames I added a comment in #2800.
- The simplest fix is to change the `if errorlevel 1 ...` to `if errorlevel 2 ...`, assuming that the `%printerror%` is wanted for Cmder logging purposes.
- Or removing [these lines](https://github.com/cmderdev/cmder/blob/59e14a17832a4ed7d9c5ac797fb0ebd0fec96b02/vendor/init.bat#L208-L210) is fine, if the `%printerror%` wasn't needed for Cmder logging.

**Comment by daxgames** [comment]
@pmsobrado we aim to please!

**Comment by DRSDavidSoft** [comment]
@pmsobrado Sorry for the message, you can blame me for the false error message. I fixed it in the master, hopefully, @chrisant996 approves

**Comment by d4k0** [comment]
ESET has released a few new versions since January that significantly improved the problem. Here are a few results from me:

cmder 1.3.24:
With exception: 1.6s
Without exception: 3,0s

cmder 1.3.21:
With exception: 1,86s
Without exception: 3.4s

It's still a bit slower, but significantly better than before. I have therefore removed the exception for `cmd.exe` again.

At work we use Trellix (they were previously called McAfee Enterprise). A few months ago, there were the same long loading times with `cmd.exe` (though not until well after January), but recently the problem has also been eliminated. I remember clink giving an error message because of the long load time.

Can anyone confirm this?

**Comment by chrisant996** [comment]
@Marco-Sulla Slowness is typically something on the host computer that is inspecting every API call, such as Antivirus or Data Loss Prevention or security tools.  This is more common on corporate managed systems, but also some Antivirus programs malfunction sometimes.

What Antivirus suites are on your machine?  Do you have ESET?  ESET's "Deep Behavioral Inspection" feature has had problems in the past where it introduced significant delays while starting programs.

**Comment by chrisant996** [comment]
Yes, that's interference from antivirus/etc.

There aren't any CMD commands between those two `echo` commands, but there is the end `)` after finishing the `if defined git_locale ( ... )` block that was just evaluated with `enabledelayedexpansion`.  That `)` triggers a bunch of system API calls and file IO for opening the .bat file again and parsing a next command, and getting/setting the console title, and a bunch of other minutia that involve system APIs.

The antivirus/etc software is intercepting one or more of the APIs and decides that it wants to do some extra processing, and the extra processing takes a bunch of time.

I agree that it doesn't make sense why that would be needed or take so long.  But it's a bug or malfunction in the antivirus/etc software -- for whatever reason, it got confused into thinking it needed to be extra work.

You'd need to identify all the antivirus and DLP and corporate monitoring tools that are installed, and try to narrow down specifically which of them are going awry, and report the issue to the respective support team for whichever software(s) are involved.  There's unfortunately nothing Cmder or ConEmu can do about this, because it's happening externally from them.

**Comment by Marco-Sulla** [comment]
I added `init.bat` to the Windows Defender exclusion list, but the start time continues to be 20s. Should I put other files or directories in exclusion list?

**Comment by Marco-Sulla** [comment]
Adding `cmd.exe` to antivirus exceptions is a backdoor opened.
I think `cmder` should wrap `init.bat` in an executable.

**Comment by Crayder** [comment]
Sorta unproductive comment... but I just wanted to mention that this magically fixed itself for me. I never did the fix above. It just started loading like normal last week. Not even updated. Nothing with cmder or clink has changed on my PC in months. My only guess is something with windows updates.

**Comment by chrisant996** [comment]
> Sorta unproductive comment... but I just wanted to mention that this magically fixed itself for me. I never did the fix above. It just started loading like normal last week. Not even updated. Nothing with cmder or clink has changed on my PC in months. My only guess is something with windows updates.

Caused/resolved by anti-virus updates.  See earlier analysis and explanations.

**Comment by SMUsamaShah** [comment]
After every single command even a simple `echo` or `cd` it was starting a 2 cmd.exe and 1 git.exe (that i could see in status bar). Meaning this was adding delay of 1-2s after every single command. Not counting the initial console startup time here.

After reading comments, I replaced my cmder-full installation with cmder-mini (which excludes git). Now there isn't no delay after running each command.

**Comment by chrisant996** [comment]
@SMUsamaShah That's interesting...

What version of Cmder are you using?

In recent versions of Cmder, git.exe gets launched _**in the background**_ and does not create any delay at all, no matter how long the git.exe command takes to run -- when the background git.exe finishes, then the prompt refreshes to show the new info.

(There are some configuration settings that can forcibly disable that, and force git.exe to run in the foreground and introduce delays.  But since you didn't mention configuring those, I'm assuming they aren't the issue.)

You could consider trying recent pre-release builds of Cmder that are not published yet.  They have some improvements, and it's been a while since an update to Cmder was published, so you might not have the improvements yet.  If you look around in recent issues you can find links e.g. from daxgames.

**Comment by SMUsamaShah** [comment]
The cmder I replaced (moved to another folder) was `Version 1.3.19.1181`.
The cmder mini I am now using is `Version 1.3.24.236`.

I tried looking for a simple way to find out why its running those processes, debug logs don't tell this info. While writing this comment I ran procmon.exe. Then did a simple `cd .` (which is an alias to `cd /d $*`). cd command completes, then does the other git related stuff.

`Process Monitor` shows cmd.exe being started with `cmd.exe /c git --no-pager config cmder.status 2>nul`, then it starts `git  --no-pager config cmder.status` which runs git.exe again with same command but using git.exe from `mingw64\bin` dir. Same cmd and git processes run once again. And then my cd command finally completes.

**Comment by SMUsamaShah** [comment]
May be it does. Because if I go to a git directory I can see the same cmd.exe and git.exe being run on the status bar but they no longer halt my command. So they are running in background as they should. So may be the packaged git and separate git installation (Program Files) was causing some weird behaviour making it run in foreground for every command, even when i wasn't in a git based dir.

**Comment by chrisant996** [comment]
> The cmder I replaced (moved to another folder) was `Version 1.3.19.1181`.
> The cmder mini I am now using is `Version 1.3.24.236`.

That was the problem.

.19 had a bug that ran part of the git stuff in the foreground by accident.

Upgrading to .24 would solve the problem, because the fix was made between those versions.

In general, when an issue is encountered in a program, it's handy to check if the latest version fixes it.

**Comment by SMUsamaShah** [comment]
Right, Thanks. I checked for update within cmder but because it only checks for conemu updates, it fooled me into thinking I have up-to-date version of cmder too. Before writing down this comment I didn't know that right way to check cmder versions is to look into its folder and see the Version file. May be there is another way but I don't know about it.
